src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1227,53 +1254,269 @@ private Type EmitFunctionCallNode(AssociativeNode node)
+                            if (p.ParameterType == typeof(char))
+                            {
+                                if (strNode.Value.Length != 1) return false;
+                            }
+                            break;
+                        case AstKind.Char:
+                            if (p.ParameterType == typeof(string)) return false;
+                            break;
+                    }
+                }
+                else if (args[argIndex] is ExprListNode exp)
+                {
+                    if (!ArrayUtils.IsEnumerable(p.ParameterType) || p.ParameterType == typeof(string))
+                    {
+                        // arg is an enumerable type, param is not.
+                        return false;
+                    }
+                    //var argRank = 0;
+                    //var arrayTypes = GetTypeStatisticsForArray(exp, ref argRank);
+                    var argRank = GetReductionDepth(exp as AssociativeNode, (x) => x is ExprListNode ? (x as ExprListNode).Exprs : null);
+                    if (argRank == -1)
+                    {
+                        // non-rectangular (jagged) array best handled by replication
+                        return false;
+                    }
+                    if (argRank != GetRank(p.ParameterType)) return false;
+                }
+                else if (args[argIndex] is IdentifierNode idn)
+                {
+                    if (idn.ReplicationGuides.Count > 0) return false;
+
+                    var t = variables[idn.Value].Item2;
+                    if (t == typeof(object)) return false;
+
+                    if (!p.ParameterType.Equals(t))
+                    {
+                        // TODO: check if rank of t matches with that of p.
+                        var argRank = GetRank(t);
+                        var paramRank = GetRank(p.ParameterType);
+                        if (argRank != paramRank) return false;
+
+                        // If both have rank 0, it could also be because their type info is ambiguous.
+                        if (argRank == 0 && paramRank == 0) return false;
+                    }
+                }
+                else return false;
+                argIndex++;
+            }
+            return true;
+        }
+
+        private static int GetRank(Type type)
+        {
+            return type.IsArray ? GetArrayRank(type) : GetEnumerableRank(type);
+        }
+
+        private static int GetArrayRank(Type type)
+        {
+            if (!type.IsArray) return 0;
+
+            return 1 + GetArrayRank(type.GetElementType());
+        }
+
+        private static int GetEnumerableRank(Type type)
+        {
+            var genericArgs = type.GetGenericArguments();
+            if (genericArgs.Length == 0) return 0;
+
+            return 1 + GetEnumerableRank(genericArgs.FirstOrDefault());
+        }
+
+        private static int GetReductionDepth<T>(T val, Func<T, IEnumerable> asArr)
```

Comment:
I can rename it, it also returns the rank of something that can be an array. In this case, I'm using it only for an expression list associative AST so I could also turn it into a function specific to finding the rank of an AST rather than making it a generic function.